DRUP-420 : Add docstore API integration to Apigee Client for PHP library

giteshk opened the pull request:
Allows the ability to interact with the Docstore apis. 
Docstore is only available on-cloud today.

giteshk committed:
Version 1 of the SpecStore API library

giteshk committed:
Version 1 of the SpecStore API library

giteshk committed:
Updates are handled through Patches

giteshk committed:
Updates are handled through Patches

giteshk committed:
Added comments to Specstore SDK classes

giteshk committed:
Added comments to Specstore SDK classes

giteshk committed:
Formatted classes correctly

giteshk committed:
Formatted classes correctly

giteshk committed:
Corrected errors in code + first test

giteshk committed:
Corrected errors in code + first test

giteshk committed:
Tests for SpecStore

giteshk committed:
Tests for SpecStore

giteshk committed:
Moved from Specstore to Docstore namespace + Added unit test cases for Foldercontroller, SpecController (partial)

giteshk committed:
Moved from Specstore to Docstore namespace + Added unit test cases for Foldercontroller, SpecController (partial)

giteshk committed:
Moved from Specstore to Docstore namespace + Added unit test cases for Foldercontroller, SpecController (partial)

giteshk committed:
Rebase from latest and apply the changes from the base

giteshk committed:
Rebase from latest and apply the changes from the base

mxr576 committed:
Do not wrap constructor parameters.

mxr576 committed:
Do not wrap constructor parameters.

mxr576 committed:
Added missing organization controller parameter.

mxr576 committed:
Added missing organization controller parameter.

mxr576 committed:
Added missing method description.

mxr576 committed:
Added missing method description.

mxr576 committed:
More clarification about how the underlying API endpoint is working.

mxr576 committed:
More clarification about how the underlying API endpoint is working.

giteshk committed:
Merge branch '2.x' of github.com:apigee/apigee-client-php into 2.x

giteshk committed:
Rename all the Docstore classes

giteshk committed:
Rename all the Docstore classes

giteshk committed:
Rename all the Docstore classes

giteshk committed:
Fix Code formatting issues

giteshk committed:
Fix Code formatting issues

giteshk committed:
Fix Code formatting issues

giteshk committed:
Rename all the Docstore classes

giteshk committed:
Rename all the Docstore classes

giteshk committed:
Added comments

giteshk committed:
Added comments

giteshk committed:
Ran php-cs-fixer one more time.

giteshk committed:
Ran php-cs-fixer one more time.

mxr576 commented on src/Api/Docstore/Serializer/DocstoreSerializer.php:
```diff
@@ -0,0 +1,38 @@
+class DocstoreSerializer extends EntitySerializer
+{
+    /**
+     * CollectionSerializer constructor.
```
Outdated doc block comment.

mxr576 commented on src/Api/Docstore/Controller/DocstoreController.php:
```diff
@@ -0,0 +1,271 @@
+/**
+ * Class DocstoreController allows for CRUD operations for Docstore entities (Folder and Doc).
+ */
+class DocstoreController extends EntityController
```
Please create an interface for this controller and move all public method definitions to that.

mxr576 commented on src/Api/Docstore/Controller/DocstoreController.php:
```diff
@@ -0,0 +1,271 @@
+            $entity->setFolder($this->getHomeFolderId());
+        }
+        $response = $this->getClient()->post(
+            ($entity instanceof Doc ? 'specs/new' : 'folders'),
```
Let's separate this logic to dedicated createFolder() and createSpec() methods. Controller methods should serve only one purpose.

mxr576 commented on src/Api/Docstore/Controller/DocstoreController.php:
```diff
@@ -0,0 +1,271 @@
+            $this->getEntitySerializer()->serialize($entity, 'json'),
+            $this->getHeaders());
+        $this->getEntitySerializer()->setPropertiesFromResponse($response, $entity);
+        $entity->setEtag($response->getHeader('ETag')[0]);
```
The entity serializer should be the one that parses _all_ information from an API response and creates/updates an entity. So etag should be set by the serializer also.

mxr576 commented on src/Api/Docstore/Controller/DocstoreController.php:
```diff
@@ -0,0 +1,271 @@
+    }
+
+    /**
+     * Upload a openAPI file to the backend.
```
s/openApi/OpenAPI/g

mxr576 commented on src/Api/Docstore/Controller/DocstoreController.php:
```diff
@@ -0,0 +1,271 @@
+     *
+     * This is what the API supports today.
+     * Uploading a JSON file and setting the content-type to application/x-yaml
+     * This will change in the future
```
Missing dot.

mxr576 commented on src/Api/Docstore/Controller/DocstoreController.php:
```diff
@@ -0,0 +1,271 @@
+    /**
+     * Get the contents of the spec.
+     *
+     * Always returns application/json
```
Missing dot.

mxr576 commented on src/Api/Docstore/Controller/DocstoreController.php:
```diff
@@ -0,0 +1,271 @@
+    /**
+     * Parses the DocStore response and generates either a Folder/Doc object.
+     */
+    private function parseDocstoreResponse(ResponseInterface $response): DocstoreObject
```
This is the task of an entity serializer and not a controller. Precisely the denormalizers inside an entity serializer should decide what object they return based on an API response. Please remove this logic here.

mxr576 commented on src/Api/Docstore/Entity/Collection.php:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Collection object in Docstore.
+ */
+class Collection extends \Apigee\Edge\Api\Docstore\Entity\DocstoreObject
```
Do not use FQCN.

mxr576 replied:
Are you sure that this is an actual entity and not just a structure?

mxr576 replied:
What could be in a collection? Folders? Docs? Can they be mixed or only from the same type? If the answer is one type per collection then I think you should introduce new FolderCollection and DocCollection classes.

giteshk replied:
Collection is the object that is returned when you pull in contents from any folder

mxr576 commented on src/Api/Docstore/Entity/Collection.php:
```diff
@@ -0,0 +1,38 @@
+ */
+class Collection extends \Apigee\Edge\Api\Docstore\Entity\DocstoreObject
+{
+    protected $kind = 'Collection';
```
As I can see there 1:1 relationship between the "kind" and the class so there is no need to store the kind as property in object - especially it should not be defined in all sub classes of the DocsStore class - rather the denormalizer could use the "kind" information from the API response to create an object with a proper type and the normalizer could add this information (property) to the API request payload based on the type of the normalized class before it gets sent to Apigee Edge. (See Rate plans as an example.)

giteshk replied:
Need this when running the create operation on the docstore entities.

Either I do this or have to set the kind in the constructor when I initialize it.

mxr576 commented on src/Api/Docstore/Entity/Doc.php:
```diff
@@ -0,0 +1,43 @@
+class Doc extends DocstoreObject
+{
+    /**
+     * url to fetch the spec file.
```
Sentence starts with a capital letter.

mxr576 commented on src/Api/Docstore/Entity/DocstoreObject.php:
```diff
@@ -0,0 +1,136 @@
+{
+    use NamePropertyAwareTrait;
+    use DescriptionPropertyAwareTrait;
+    use \Apigee\Edge\Entity\CommonEntityPropertiesAwareTrait;
```
Do not use FQCN.

mxr576 commented on src/Api/Docstore/Entity/DocstoreObject.php:
```diff
@@ -0,0 +1,136 @@
+    use DescriptionPropertyAwareTrait;
+    use \Apigee\Edge\Entity\CommonEntityPropertiesAwareTrait;
+
+    protected $permissions;
```
Missing typehint from properties, some of them may deserve a description as well.

mxr576 commented on src/Api/Docstore/Entity/DocstoreObject.php:
```diff
@@ -0,0 +1,136 @@
+     *
+     * @return mixed
+     */
+    public function getIsTrashed()
```
This is a boolean property so its getter should `isTranshed()` and it setter should be `setTrashed()`. 
See Address' name converter as an example for how you can fix this:
https://github.com/apigee/apigee-client-php/blob/2.x/src/Api/Monetization/NameConverter/AddressNameConverter.php#L31
https://github.com/apigee/apigee-client-php/blob/2.x/src/Api/Monetization/Denormalizer/AddressDenormalizer.php#L41
https://github.com/apigee/apigee-client-php/blob/2.x/src/Api/Monetization/Normalizer/AddressNormalizer.php#L41
https://github.com/apigee/apigee-client-php/blob/2.x/src/Api/Monetization/Structure/Address.php#L29

giteshk replied:
I implemented this. 

BTW the trashed flag will always return false. There is no concept of Trash in docstore as it stands today.  I think they might implement this in the future.

mxr576 commented on src/Api/Docstore/Entity/DocstoreObject.php:
```diff
@@ -0,0 +1,136 @@
+/**
+ * Abstract object to represent the Docstore entities.
+ */
+abstract class DocstoreObject extends Entity
```
Please define an interface for this class, move all its public definition to that class and please define interfaces for all its subclasses as well.

mxr576 commented on tests/Api/Docstore/Controller/DocstoreControllerTest.php:
```diff
@@ -0,0 +1,664 @@
+ * Class DocstoreControllerTest.
+ *
+ * @group controller
+ * @group management
```
This test does not belong to a management API test, this is what this group stands for. I'd rather introduce a new `docstore`.

mxr576 commented on tests/Api/Docstore/Controller/DocstoreControllerTest.php:
```diff
@@ -0,0 +1,664 @@
+        $spec = new Doc(['name' => static::getSampleDocName()]);
+        static::entityController()->create($spec);
+
+        $homeFolder = static::entityController()->load('/homeFolder');
```
Why `/` is necessary? It should not be. `load()` method on an entity controller loads an entity by id.

giteshk replied:
entity id's for all docstore entities begin with a "/"

mxr576 commented on tests/Api/Docstore/Controller/DocstoreControllerTest.php:
```diff
@@ -0,0 +1,664 @@
+    public static function tearDownAfterClass(): void
+    {
+        parent::tearDownAfterClass();
+        static::removeTestData();
```
Feel free to inline the method content's here. It seems that you only use this in one place.

mxr576 commented on src/Api/Docstore/Controller/DocstoreController.php:
```diff
@@ -0,0 +1,271 @@
+     */
+    public function update(DocstoreObject $entity): void
+    {
+        $uri = $this->getEntityEndpointUri($entity->id());
```
Unnecessary new variable `$uri`, you only use it in one place.

mxr576 commented on src/Api/Docstore/Controller/DocstoreController.php:
```diff
@@ -0,0 +1,271 @@
+    public function update(DocstoreObject $entity): void
+    {
+        $uri = $this->getEntityEndpointUri($entity->id());
+        $update_arr = [];
```
Use camelCase variable names inside the PHP API Client.

mxr576 commented on src/Api/Docstore/Controller/DocstoreController.php:
```diff
@@ -0,0 +1,271 @@
+        // Update an existing entity.
+        $response = $this->getClient()->patch(
+            $uri,
+            $this->getEntitySerializer()->serialize($update_arr, 'json'),
```
Can't you pass the entity directly to the serializer that should generate the required request payload from it? If not, why?

giteshk replied:
The patch method only takes certain values. I tried passing the entire entity and it fails.

mxr576 commented on src/Api/Docstore/Controller/DocstoreController.php:
```diff
@@ -0,0 +1,271 @@
+    public function getSpecContents(Doc $entity): string
+    {
+        $response = $this->getClient()->get($entity->getContent(),
+            $this->getHeaders() + ['Accept' => 'application/json, text/plain, */*']);
```
Isn't only "application/json" what this API can return? If not "text/plan" could be acceptable if the API returns a JSON file with incorrect content type, but I do not understand why "*/*" is necessary. Basically it means we accept anything, although our serializer can only read JSON.

giteshk replied:
This is the actual openAPI specification that is uploaded in edge.

The API documentation had that in the instructions. But I have tested this with application/json and my tests ran. So I am setting this to only be application/json

mxr576 commented:
No offense, but I do not get how this code passed on the code style to be honest, probably the code style checker requires some adjustments ;)
* variable names should be camelCase
* At least variable type hint should be added to the class properties, one new line between each property.
* All entity and controller classes should have an interface defined that describes its public behaviors - this is something that we can not validate with static code analyses at this moment.

mxr576 commented on src/Api/Docstore/Controller/DocstoreController.php:
```diff
@@ -0,0 +1,271 @@
+    }
+
+    /**
+     * Docstore entity id start with a "/".
```
Entity ids should not start with `/`. Just add that to the path builder...
https://github.com/apigee/apigee-client-php/blob/8017c3993ab78c239744b4b1fdd25deebf61678b/src/Api/Management/Controller/DeveloperAppCredentialController.php#L67

giteshk replied:
Entity ID returned by the backend API is in this format "/c3Rvc-ZG9j-93459"

The logic to generate the entity URL is abstracted in this one method.

mxr576 committed:
Merge pull request #32 from mxr576/company_api

Add missing organization controller parameter

mxr576 committed:
Bumped client version.

mxr576 committed:
Bumped client version.

mxr576 committed:
Added changes since 2.0.0-alpha6.

mxr576 committed:
Added changes since 2.0.0-alpha6.

mxr576 committed:
Merge pull request #35 from mxr576/2.0.0-prep

2.0.0 release

giteshk committed:
variable names fixed to camelCase
Created Interface for the Controller

giteshk committed:
variable names fixed to camelCase
Created Interface for the Controller

giteshk committed:
variable names fixed to camelCase
Created Interface for the Controller

giteshk committed:
variable names fixed to camelCase
Created Interface for the Controller and Entity
Renamed DocstoreObject to DocstoreEntity
Added variable type hints

giteshk committed:
variable names fixed to camelCase
Created Interface for the Controller and Entity
Renamed DocstoreObject to DocstoreEntity
Added variable type hints

giteshk committed:
variable names fixed to camelCase
Created Interface for the Controller and Entity
Renamed DocstoreObject to DocstoreEntity
Added variable type hints

giteshk committed:
Fix psalm errors

giteshk committed:
Fix psalm errors

giteshk committed:
Merge remote-tracking branch 'origin/2.x' into 2.x

giteshk committed:
1. Created a NameConverter for mapping isTrashed to trashed flag
2. Re-packaged the NameConverterBase class

giteshk committed:
Corrected Psalm errors

giteshk commented:
I think there was a merge issue. I will close this and create a new pull request.
